File: conformance/run-1/components/channel-wiring.py
import asyncio

from jiggy.flow import OperationError, handle


async def run(context):
    value = context.input
    if value["role"] == "root":
        source = await context.channel(contract="./updates.json")
        incompatible = False
        try:
            await context.run_child_flow(
                operation_id="incompatible", slot="incompatible", input=None,
                channels={"events": source.send},
            )
        except OperationError as error:
            if error.code != "INVALID_INPUT":
                raise
            incompatible = True
        channels = {"events": source.receive}
        if "progress" in context.channels:
            channels["progress"] = context.channels["progress"]
        work, monitor = await asyncio.gather(
            context.run_child_flow(operation_id="worker", slot="worker", input={"role": "worker"},
                                   channels={"events": source.send}),
            context.run_child_flow(operation_id="monitor", slot="monitor",
                                   input={"role": "monitor", "stop": value.get("stop", False)},
                                   channels=channels),
            return_exceptions=True,
        )
        if isinstance(work, BaseException):
            raise work
        if isinstance(monitor, BaseException):
            raise monitor
        return {"outcome": "done", "output": {"incompatible": incompatible, "work": work, "monitor": monitor}}
    if value["role"] == "worker":
        result = await context.call_capability(
            operation_id="answer", slot="agent", method="run", input=None,
            channels={"events": context.channels["events"]},
        )
        return {"outcome": "done", "output": result}
    if value["role"] != "monitor":
        raise TypeError("unknown role")
    events = context.channels["events"]
    progress = context.channels.get("progress")
    selected = []
    complete = True
    try:
        async with events:
            async for item in events:
                if not isinstance(item, dict) or "text" not in item:
                    continue
                selected.append(item["text"])
                if progress is not None:
                    await progress.send(item["text"])
                if value.get("stop", False):
                    complete = False
                    break
    except OperationError as error:
        if error.code != "LAGGED":
            raise
        complete = False
    if progress is not None:
        await progress.close()
    return {"outcome": "done", "output": {"complete": complete, "selected": selected}}


handle(run)

File: conformance/run-1/components/channel-wiring.ts
import {
  type ChannelEndpoint,
  handle,
  type JsonValue,
  OperationError,
  type RunResult,
} from '../../../packages/flow-sdk/src/index'

await handle(async (run): Promise<RunResult> => {
  const input = run.input
  if (!input || typeof input !== 'object' || !('role' in input))
    throw new TypeError('role required')
  if (input.role === 'root') {
    const source = await run.channel({ contract: './updates.json' })
    let incompatible = false
    try {
      await run.call({
        operationId: 'incompatible',
        slot: 'incompatible',
        input: null,
        channels: { events: source.send },
      })
    } catch (error) {
      if (!(error instanceof OperationError) || error.code !== 'INVALID_INPUT') throw error
      incompatible = true
    }
    const work = run.call({
      operationId: 'worker',
      slot: 'worker',
      input: { role: 'worker' },
      channels: { events: source.send },
    })
    const channels: Record<string, ChannelEndpoint> = { events: source.receive }
    if (run.channels.progress) channels.progress = run.channels.progress
    const monitor = run.call({
      operationId: 'monitor',
      slot: 'monitor',
      input: { role: 'monitor', stop: input.stop === true },
      channels,
    })
    const [worked, watched] = await Promise.allSettled([work, monitor])
    if (worked.status === 'rejected') throw worked.reason
    if (watched.status === 'rejected') throw watched.reason
    return { outcome: 'done', output: { incompatible, work: worked.value, monitor: watched.value } }
  }
  if (input.role === 'worker') {
    const result = await run.call({
      operationId: 'answer',
      slot: 'agent',
      input: null,
      channels: { events: run.channels.events! },
    })
    return result
  }
  if (input.role !== 'monitor') throw new TypeError('unknown role')
  const events = run.channels.events
  if (!events || events.direction !== 'receive') throw new TypeError('events receiver required')
  const progress = run.channels.progress
  if (progress && progress.direction !== 'send') throw new TypeError('progress sender required')
  const selected: JsonValue[] = []
  let complete = true
  try {
    for await (const value of events) {
      if (!value || typeof value !== 'object' || !('text' in value)) continue
      selected.push(value.text)
      if (progress) await progress.send(value.text)
      if (input.stop === true) {
        complete = false
        break
      }
    }
  } catch (error) {
    if (!(error instanceof OperationError) || error.code !== 'LAGGED') throw error
    complete = false
  } finally {
    await events.close()
  }
  if (progress) await progress.close()
  return { outcome: 'done', output: { complete, selected } }
})
